plugins/woocommerce/includes/wc-cart-functions.php

```diff
@@ -170,6 +170,9 @@ function wc_format_list_of_items( $items ) {
 function wc_clear_cart_after_payment() {
 	global $wp;
 
+	$should_clear_cart_after_payment = false;
```

Comment:
**(1/4)** Thanks for your patience while we worked through this. 

I'd like to propose a final amendment, after which we can probably go ahead and merge, but let me know if this would be problematic in any way (cc @vedanshujain, I'd appreciate your eyes on this, as I may be being over-cautious).

In outline, here's what I'm thinking. Before this change, the function worked something like this:

- Run on every `template_redirect` ⚠️ 
- If `order-received` and we have a valid order instance, clear the cart.
- If we have an order awaiting payment, and the status is no longer `pending`, `failed` or `cancelled`, also clear the cart.

That was pretty efficient ... but now we have the following: 

- Run on every `template_redirect` ⚠️ 
- If `order-received` and we have a valid order instance, consider clearing the cart.
- Or, if we have an order awaiting payment, and the status is no longer `pending`, `failed` or `cancelled`, also consider clearing the cart.
- Take our decision, let third parties filter it, then actually clear the cart or not.

It's definitely not awful, but because this is running on every single `template_redirect` (hence my little ⚠️ icons), I'm wondering if we can introduce some extra efficiency gains, without too much work:

```suggestion
	$should_clear_cart_after_payment = false;
	$after_payment                   = false;
```

This new var tracks if we think a payment just took place. See next comment.

Reply:
I think we can proceed without (Vedanshu probably won't be able to look at this soon), and we will have some space to correct if necessary as we're at a relatively early stage within the 9.3 cycle.

Reply:
I've gone ahead and made those changes. Let me know what you think.